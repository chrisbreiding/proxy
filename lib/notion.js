const rp = require('request-promise')

const getData = ({ notionPageId, notionToken }) => {
  return rp({
    uri: `https://api.notion.com/v1/blocks/${notionPageId}/children`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

module.exports = {
  getData,
}
