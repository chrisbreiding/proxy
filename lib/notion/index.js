const { addUpcomingWeek } = require('./upcoming-week')
const { getData } = require('./data')
const { onSocket } = require('./shopping')

module.exports = {
  addUpcomingWeek,
  getData,
  onSocket,
}
