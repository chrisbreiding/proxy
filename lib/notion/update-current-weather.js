const debug = require('debug')('proxy:scripts')

const { getData: getWeatherData } = require('../weather')
const { updateBlock, makeTextPart, getWeatherIcon } = require('./util')
const { compact } = require('../util/collections')
const { getEnv } = require('../util/env')

const notionToken = getEnv('NOTION_TOKEN')
const currentWeatherId = getEnv('CURRENT_WEATHER_ID')
const weatherLocation = getEnv('WEATHER_LOCATION')

debug('ENV:', {
  notionToken,
  currentWeatherId,
  weatherLocation,
})

const getWeather = async () => {
  const weather = await getWeatherData({ location: weatherLocation })

  return weather.currently
}

const makePrecipPart = (condition, info) => {
  return condition ? makeTextPart(`(${info}) `, 'gray') : undefined
}

const getColor = (temp) => {
  if (temp <= 32) return 'purple'
  if (temp > 32 && temp <= 45) return 'blue'
  if (temp > 45 && temp <= 75) return 'green'
  if (temp > 75 && temp <= 90) return 'orange'
  return 'red'
}

const updateBlockWeather = async (weather) => {
  const content = {
    heading_1: {
      text: compact([
        makeTextPart(`${getWeatherIcon(weather.icon)} `),
        makePrecipPart(weather.icon === 'rain' && weather.precipProbability >= 0.01, `${Math.round(weather.precipProbability * 100)}%`),
        makePrecipPart(weather.icon === 'snow' && weather.precipAccumulation >= 0.1, `${weather.precipAccumulation.toFixed(2)}in`),
        makeTextPart(`${Math.round(weather.temperature)}Â°`, getColor(weather.temperature)),
      ]),
    },
  }

  const newText = content.heading_1.text
  .map(({ text }) => text.content)
  .join('')
  .trim()

  // eslint-disable-next-line no-console
  console.log(`Update current weather to "${newText}"`)

  await updateBlock({ notionToken, block: content, blockId: currentWeatherId })
}

const updateWeather = async () => {
  try {
    const weather = await getWeather({ location: weatherLocation })

    await updateBlockWeather(weather)

    // eslint-disable-next-line no-console
    console.log('Successfully updated current weather')

    process.exit(0)
  } catch (error) {
    // eslint-disable-next-line no-console
    console.log('Updating current weather failed:')
    // eslint-disable-next-line no-console
    console.log(error.stack)

    process.exit(1)
  }
}

updateWeather()
