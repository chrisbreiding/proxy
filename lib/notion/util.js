const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
