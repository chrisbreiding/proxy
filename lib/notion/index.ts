import { addUpcomingWeek, upcomingWeekView } from './upcoming-week'
import { getBlockChildren } from './util'
import { onSocket } from './shopping'

interface GetDataOptions {
  notionToken: string
  notionPageId: string
}

export function getNotionData ({ notionToken, notionPageId }: GetDataOptions) {
  return getBlockChildren({ notionToken, pageId: notionPageId })
}

export {
  addUpcomingWeek,
  upcomingWeekView,
  onSocket,
}
