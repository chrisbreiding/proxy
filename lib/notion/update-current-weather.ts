import dayjs from 'dayjs'

import {
  updateBlock,
  makeTextPart,
  getBlockChildren,
  deleteBlock,
  OwnBlock,
  appendBlockChildrenDeep,
} from './util'
import { compact } from '../util/collections'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { DayWeather, getWeatherData, getWeatherIcon } from '../weather'
import { makeConditionParts, makePrecipPart, makeTemperatureParts } from './weather'

function getColor (temp: number) {
  if (temp <= 32) return 'purple'
  if (temp > 32 && temp <= 45) return 'blue'
  if (temp > 45 && temp <= 75) return 'green'
  if (temp > 75 && temp <= 90) return 'orange'
  return 'red'
}

function makeTableCells (weather: DayWeather[]) {
  return weather.map((dayWeather) => {
    const date = dayjs.unix(dayWeather.time).format('ddd, M/D')

    return {
      type: 'table_row',
      content: {
        cells: [
          [makeTextPart(`${date}`)], // Mon, 11/1
          [...makeTemperatureParts(dayWeather)], // 33° / 62°
          [...makeConditionParts(dayWeather)], // ☀️ | ☔️ (82%)
        ],
      },
    } as OwnBlock
  })
}

interface UpdateTableWeatherOptions {
  notionToken: string
  currentWeatherId: string
  weather: DayWeather[]
}

async function updateTableWeather ({ notionToken, currentWeatherId, weather }: UpdateTableWeatherOptions) {
  const children = await getBlockChildren({ notionToken, pageId: currentWeatherId })

  if (children.length) {
    deleteBlock({ blockId: children[0].id, notionToken })
  }

  const table = {
    type: 'table',
    content: {
      table_width: 3,
      has_column_header: false,
      has_row_header: false,
    },
    children: makeTableCells(weather),
  } as OwnBlock

  await appendBlockChildrenDeep({ blocks: [table], notionToken, pageId: currentWeatherId })
}

interface UpdateWeatherOptions {
  currentWeatherId: string
  location: string
  notionToken: string
}

export async function updateWeather ({ currentWeatherId, location, notionToken }: UpdateWeatherOptions) {
  const weather = await getWeatherData(location)

  const {
    icon,
    precipProbability,
    precipAccumulation,
    temperature,
  } = weather.currently
  const roundedTemperature = Math.round(temperature)

  const block = {
    type: 'heading_1' as const,
    content: {
      rich_text: compact([
        makeTextPart(`${getWeatherIcon(icon)} `),
        makePrecipPart(icon === 'rain' && precipProbability >= 0.01, `${Math.round(precipProbability * 100)}%`),
        makePrecipPart(icon === 'snow' && precipAccumulation >= 0.1, `${(precipAccumulation || 0).toFixed(2)}in`),
        makeTextPart(`${roundedTemperature}°`, getColor(roundedTemperature)),
      ]),
      color: 'default' as const,
      is_toggleable: true,
    },
  }

  const newText = block.content.rich_text
  .map(({ text }) => text.content)
  .join('')
  .trim()

  debug(`Update current weather to '${newText}'`)

  await updateBlock({ notionToken, block, blockId: currentWeatherId })
  await updateTableWeather({ currentWeatherId, notionToken, weather: weather.daily.data })
}

export default async function main () {
  const currentWeatherId = getEnv('CURRENT_WEATHER_ID')!
  const location = getEnv('WEATHER_LOCATION')!
  const notionToken = getEnv('NOTION_TOKEN')!

  debugVerbose('ENV:', {
    currentWeatherId,
    location,
    notionToken,
  })

  try {
    debug('Updating current weather...')

    await updateWeather({
      currentWeatherId,
      location,
      notionToken,
    })
  } catch (error: any) {
    debug('Updating current weather failed:')
    debug(error?.stack || error)
  }
}
