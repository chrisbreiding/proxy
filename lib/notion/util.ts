import type { Method } from 'axios'
import dayjs from 'dayjs'
import type {
  ApiColor,
  BlockObjectResponse,
  BookmarkBlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ChildDatabaseBlockObjectResponse,
  ChildPageBlockObjectResponse,
  ColumnBlockObjectResponse,
  ColumnListBlockObjectResponse,
  DividerBlockObjectResponse,
  EmbedBlockObjectResponse,
  EquationBlockObjectResponse,
  FileBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ImageBlockObjectResponse,
  LinkPreviewBlockObjectResponse,
  LinkToPageBlockObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  NumberedListItemBlockObjectResponse,
  ParagraphBlockObjectResponse,
  PdfBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  SyncedBlockBlockObjectResponse,
  TableBlockObjectResponse,
  TableOfContentsBlockObjectResponse,
  TableRowBlockObjectResponse,
  TemplateBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
  UnsupportedBlockObjectResponse,
  UpdateBlockBodyParameters,
  VideoBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints'

import { clone, compact } from '../util/collections'
import { debug } from '../util/debug'
import { request } from '../util/network'

export const dateRegex = /(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat), (\d{1,2}\/\d{1,2})/

export function getDateFromText (dateText: string) {
  const currentDate = dayjs()
  // originally assume the date's year matches the current year
  const date = dayjs(`${dateText}/${currentDate.year()}`, 'M/D/YYYY')
  // if the current month is after the date's month, we've crossed over years
  // and the date's year should be the next year. usually, it's because it's
  // currently December, but the date's month is January. this can be
  // generically applied, so it will work if it's currently June, but we're
  // trying to get the weather for April. we don't want past weather, only
  // future weather, so assume it's for the following year.
  if (currentDate.month() > date.month()) {
    return date.add(1, 'year')
  }

  return date
}

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({ notionToken, path: `blocks/${pageId}/children` }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

type Content = BookmarkBlockObjectResponse['bookmark']
| BulletedListItemBlockObjectResponse['bulleted_list_item']
| CalloutBlockObjectResponse['callout']
| ChildDatabaseBlockObjectResponse['child_database']
| ChildPageBlockObjectResponse['child_page']
| ColumnBlockObjectResponse['column']
| ColumnListBlockObjectResponse['column_list']
| DividerBlockObjectResponse['divider']
| EmbedBlockObjectResponse['embed']
| EquationBlockObjectResponse['equation']
| FileBlockObjectResponse['file']
| Heading1BlockObjectResponse['heading_1']
| Heading2BlockObjectResponse['heading_2']
| Heading3BlockObjectResponse['heading_3']
| ImageBlockObjectResponse['image']
| LinkPreviewBlockObjectResponse['link_preview']
| LinkToPageBlockObjectResponse['link_to_page']
| NumberedListItemBlockObjectResponse['numbered_list_item']
| ParagraphBlockObjectResponse['paragraph']
| PdfBlockObjectResponse['pdf']
| QuoteBlockObjectResponse['quote']
| SyncedBlockBlockObjectResponse['synced_block']
| TableBlockObjectResponse['table']
| TableOfContentsBlockObjectResponse['table_of_contents']
| TableRowBlockObjectResponse['table_row']
| TemplateBlockObjectResponse['template']
| ToDoBlockObjectResponse['to_do']
| ToggleBlockObjectResponse['toggle']
| UnsupportedBlockObjectResponse['unsupported']
| VideoBlockObjectResponse['video']

export interface OwnBlock {
  children?: OwnBlock[] | NotionBlock[]
  type: BlockObjectResponse['type']
  content: Content
}

export interface NotionBlock extends OwnBlock {
  children?: NotionBlock[]
  has_children: boolean
  id: string
}

export type Block = OwnBlock | NotionBlock

function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
}

interface MakeAppendRequestOptions {
  blocks: OutgoingBlock[]
  notionToken: string
  pageId: string
}

export function makeAppendRequest ({ notionToken, pageId, blocks }: MakeAppendRequestOptions) {
  return makeRequest<ListBlockChildrenResponse>({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface AppendBlockChildrenOptions {
  blocks: OwnBlock[]
  notionToken: string
  pageId: string
}

export async function appendBlockChildrenDeep ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  function moveChildren (blocks: NotionBlock[] | OwnBlock[]) {
    return blocks.map((block) => {
      const convertedBlock = convertBlockToOutgoingBlock(block)

      if (!block.children) return convertedBlock

      // @ts-ignore
      convertedBlock[block.type].children = moveChildren(block.children)

      return convertedBlock
    })
  }

  return makeAppendRequest({ notionToken, pageId, blocks: moveChildren(blocks) })
}

export async function appendBlockChildren ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  let toAppend: OutgoingBlock[] = []

  for (const block of blocks) {
    const convertedBlock = convertBlockToOutgoingBlock(block)

    if (block.children) {
      const children = block.children

      toAppend.push(convertedBlock)

      const { results } = await makeAppendRequest({ notionToken, pageId, blocks: toAppend })
      toAppend = []

      await appendBlockChildren({
        notionToken,
        pageId: results[results.length - 1].id,
        blocks: children,
      })
    } else {
      toAppend.push(convertedBlock)
    }
  }

  if (toAppend.length) {
    return makeAppendRequest({
      notionToken,
      pageId,
      blocks: toAppend,
    })
  }
}

interface UpdateBlockOptions {
  notionToken: string
  blockId: string
  block: Block
}

export function updateBlock ({ notionToken, blockId, block }: UpdateBlockOptions) {
  const convertedBlock = convertBlockToOutgoingBlock(block)

  return makeRequest<void>({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: convertedBlock,
  })
}

interface UpdatePageOptions {
  notionToken: string
  pageId: string
  properties: {
    [key: string]: {
      date: {
        start: string
      }
    }
  }
}

export function updatePage ({ notionToken, pageId, properties }: UpdatePageOptions) {
  return makeRequest<void>({
    notionToken,
    body: { properties },
    method: 'patch',
    path: `pages/${pageId}`,
  })
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
}

export function queryDatabases ({ notionToken, databaseId }: QueryDatabasesOptions) {
  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
  })
}

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type

  return {
    // @ts-ignore
    content: clone(block[type] as Content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

type OutgoingBlock = UpdateBlockBodyParameters & {
  object: string
  type: BlockObjectResponse['type']
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  text: string
  type: BlockObjectResponse['type']
}

export function makeBlock ({ text, type = 'paragraph', annotations, children }: MakeBlockOptions) {
  return {
    type,
    children,
    content: {
      rich_text: [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ],
    },
  } as OwnBlock
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

export function getMonthNameFromIndex (monthIndex: number, short = false) {
  return dayjs().month(monthIndex).format(short ? 'MMM' : 'MMMM')
}

export function getMonths ({ short }: { short?: boolean } = {}) {
  return Array.from(new Array(12)).map((_, i) => {
    return getMonthNameFromIndex(i, short)
  })
}

/* c8 ignore start */
const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
/* c8 ignore stop */
