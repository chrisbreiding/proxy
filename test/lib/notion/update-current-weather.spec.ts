import nock from 'nock'
import { describe, it } from 'vitest'

process.env.DARK_SKY_API_KEY = 'dark-sky-key'

import { updateWeather } from '../../../lib/notion/update-current-weather'
import type { WeatherIcon } from '../../../lib/weather'
import { snapshotBody } from '../../util'
import {
  nockAppendBlockChildren,
  nockGetBlockChildren,
  nockNotion,
  nockUpdateBlock,
  notionFixture as fixture,
  notionFixtureContents as fixtureContents,
} from './util'

interface WeatherProps {
  temperature?: number
  icon?: WeatherIcon
  precipProbability?: number
  precipAccumulation?: number
}

describe('lib/notion/update-current-weather', () => {
  function nockCurrentWeather (props: WeatherProps) {
    const weather = fixtureContents('weather/weather')

    weather.currently = {
      ...weather.currently,
      ...props,
    }

    nock('https://api.darksky.net')
    .get('/forecast/dark-sky-key/lat,lng?exclude=minutely,flags&extend=hourly')
    .reply(200, weather)

    nockGetBlockChildren('current-weather-id', { fixture: 'weather/current-weather-children' })
    nockNotion({ path: '/v1/blocks/table-id', method: 'delete' })
    nockAppendBlockChildren({ id: 'current-weather-id' })

    return snapshotBody(nockUpdateBlock('current-weather-id'))
  }

  it('updates the current weather block with temperature and conditions', async () => {
    nock('https://api.darksky.net')
    .get('/forecast/dark-sky-key/lat,lng?exclude=minutely,flags&extend=hourly')
    .replyWithFile(200, fixture('weather/weather'), {
      'Content-Type': 'application/json',
    })

    nockGetBlockChildren('current-weather-id', { fixture: 'weather/current-weather-children' })
    nockNotion({ path: '/v1/blocks/table-id', method: 'delete' })

    const snapshots = [
      snapshotBody(nockUpdateBlock('current-weather-id'), 'heading-update'),
      snapshotBody(nockAppendBlockChildren({ id: 'current-weather-id' }), 'table-append'),
    ]

    await updateWeather({
      notionToken: 'notion-token',
      currentWeatherId: 'current-weather-id',
      location: 'lat,lng',
    })

    await Promise.all(snapshots)
  })

  it('colors cool temperature blue', async () => {
    const snapshot = nockCurrentWeather({ temperature: 40 })

    await updateWeather({
      notionToken: 'notion-token',
      currentWeatherId: 'current-weather-id',
      location: 'lat,lng',
    })

    await snapshot
  })

  it('colors warm temperature green', async () => {
    const snapshot = nockCurrentWeather({ temperature: 60 })

    await updateWeather({
      notionToken: 'notion-token',
      currentWeatherId: 'current-weather-id',
      location: 'lat,lng',
    })

    await snapshot
  })

  it('colors hot temperature orange', async () => {
    const snapshot = nockCurrentWeather({ temperature: 85 })

    await updateWeather({
      notionToken: 'notion-token',
      currentWeatherId: 'current-weather-id',
      location: 'lat,lng',
    })

    await snapshot
  })

  it('colors sweltering temperature red', async () => {
    const snapshot = nockCurrentWeather({ temperature: 100 })

    await updateWeather({
      notionToken: 'notion-token',
      currentWeatherId: 'current-weather-id',
      location: 'lat,lng',
    })

    await snapshot
  })

  it('displays rain icon and probability if rain', async () => {
    const snapshot = nockCurrentWeather({
      icon: 'rain',
      precipProbability: 0.45,
    })

    await updateWeather({
      notionToken: 'notion-token',
      currentWeatherId: 'current-weather-id',
      location: 'lat,lng',
    })

    await snapshot
  })

  it('displays snow icon and accumulation if snow', async () => {
    const snapshot = nockCurrentWeather({
      icon: 'snow',
      precipAccumulation: 2,
    })

    await updateWeather({
      notionToken: 'notion-token',
      currentWeatherId: 'current-weather-id',
      location: 'lat,lng',
    })

    await snapshot
  })
})
