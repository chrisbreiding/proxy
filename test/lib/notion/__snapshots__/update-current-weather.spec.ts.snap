// Vitest Snapshot v1

exports[`lib/notion/update-current-weather > colors cool temperature blue 1`] = `
"# 🌑 40° {💙}
| Wed, 12/28 | 23° {💙} / {🩶} 62° {🧡} | 🌪  |
| ---------- | ------------------------ | --- |
| Thu, 12/29 | 45° {💙} / {🩶} 66° {🧡} | ⛅️  |
| ---------- | ------------------------ | --- |
| Fri, 12/30 | 43° {💙} / {🩶} 66° {🧡} | 💨  |
| ---------- | ------------------------ | --- |
| Sat, 12/31 | 33° {💙} / {🩶} 54° {🧡} | ☔️ (85%) {🩶}  |
| ---------- | ------------------------ | -------------- |
| Sun, 1/1 | 25° {💙} / {🩶} 51° {🧡} | ❄️  |
| -------- | ------------------------ | --- |
| Mon, 1/2 | 26° {💙} / {🩶} 51° {🧡} | 🌫  |
| -------- | ------------------------ | --- |
| Tue, 1/3 | 29° {💙} / {🩶} 63° {🧡} | 🥵  |
| -------- | ------------------------ | --- |
| Wed, 1/4 | 28° {💙} / {🩶} 44° {🧡} | 🌀  |
| -------- | ------------------------ | --- |
Updated Dec 28, 12:00am {🩶}
"
`;

exports[`lib/notion/update-current-weather > colors hot temperature orange 1`] = `
"# ☀️ 85° {🧡}
| Wed, 12/28 | 23° {💙} / {🩶} 62° {🧡} | 🌪  |
| ---------- | ------------------------ | --- |
| Thu, 12/29 | 45° {💙} / {🩶} 66° {🧡} | ⛅️  |
| ---------- | ------------------------ | --- |
| Fri, 12/30 | 43° {💙} / {🩶} 66° {🧡} | 💨  |
| ---------- | ------------------------ | --- |
| Sat, 12/31 | 33° {💙} / {🩶} 54° {🧡} | ☔️ (85%) {🩶}  |
| ---------- | ------------------------ | -------------- |
| Sun, 1/1 | 25° {💙} / {🩶} 51° {🧡} | ❄️  |
| -------- | ------------------------ | --- |
| Mon, 1/2 | 26° {💙} / {🩶} 51° {🧡} | 🌫  |
| -------- | ------------------------ | --- |
| Tue, 1/3 | 29° {💙} / {🩶} 63° {🧡} | 🥵  |
| -------- | ------------------------ | --- |
| Wed, 1/4 | 28° {💙} / {🩶} 44° {🧡} | 🌀  |
| -------- | ------------------------ | --- |
Updated Dec 28, 12:00am {🩶}
"
`;

exports[`lib/notion/update-current-weather > colors sweltering temperature red 1`] = `
"# 🌨 100° {💔}
| Wed, 12/28 | 23° {💙} / {🩶} 62° {🧡} | 🌪  |
| ---------- | ------------------------ | --- |
| Thu, 12/29 | 45° {💙} / {🩶} 66° {🧡} | ⛅️  |
| ---------- | ------------------------ | --- |
| Fri, 12/30 | 43° {💙} / {🩶} 66° {🧡} | 💨  |
| ---------- | ------------------------ | --- |
| Sat, 12/31 | 33° {💙} / {🩶} 54° {🧡} | ☔️ (85%) {🩶}  |
| ---------- | ------------------------ | -------------- |
| Sun, 1/1 | 25° {💙} / {🩶} 51° {🧡} | ❄️  |
| -------- | ------------------------ | --- |
| Mon, 1/2 | 26° {💙} / {🩶} 51° {🧡} | 🌫  |
| -------- | ------------------------ | --- |
| Tue, 1/3 | 29° {💙} / {🩶} 63° {🧡} | 🥵  |
| -------- | ------------------------ | --- |
| Wed, 1/4 | 28° {💙} / {🩶} 44° {🧡} | 🌀  |
| -------- | ------------------------ | --- |
Updated Dec 28, 12:00am {🩶}
"
`;

exports[`lib/notion/update-current-weather > colors warm temperature green 1`] = `
"# ☁️ 60° {💚}
| Wed, 12/28 | 23° {💙} / {🩶} 62° {🧡} | 🌪  |
| ---------- | ------------------------ | --- |
| Thu, 12/29 | 45° {💙} / {🩶} 66° {🧡} | ⛅️  |
| ---------- | ------------------------ | --- |
| Fri, 12/30 | 43° {💙} / {🩶} 66° {🧡} | 💨  |
| ---------- | ------------------------ | --- |
| Sat, 12/31 | 33° {💙} / {🩶} 54° {🧡} | ☔️ (85%) {🩶}  |
| ---------- | ------------------------ | -------------- |
| Sun, 1/1 | 25° {💙} / {🩶} 51° {🧡} | ❄️  |
| -------- | ------------------------ | --- |
| Mon, 1/2 | 26° {💙} / {🩶} 51° {🧡} | 🌫  |
| -------- | ------------------------ | --- |
| Tue, 1/3 | 29° {💙} / {🩶} 63° {🧡} | 🥵  |
| -------- | ------------------------ | --- |
| Wed, 1/4 | 28° {💙} / {🩶} 44° {🧡} | 🌀  |
| -------- | ------------------------ | --- |
Updated Dec 28, 12:00am {🩶}
"
`;

exports[`lib/notion/update-current-weather > displays rain icon and probability if rain 1`] = `
"# ☔️ (45%) {🩶} 23° {💜}
| Wed, 12/28 | 23° {💙} / {🩶} 62° {🧡} | 🌪  |
| ---------- | ------------------------ | --- |
| Thu, 12/29 | 45° {💙} / {🩶} 66° {🧡} | ⛅️  |
| ---------- | ------------------------ | --- |
| Fri, 12/30 | 43° {💙} / {🩶} 66° {🧡} | 💨  |
| ---------- | ------------------------ | --- |
| Sat, 12/31 | 33° {💙} / {🩶} 54° {🧡} | ☔️ (85%) {🩶}  |
| ---------- | ------------------------ | -------------- |
| Sun, 1/1 | 25° {💙} / {🩶} 51° {🧡} | ❄️  |
| -------- | ------------------------ | --- |
| Mon, 1/2 | 26° {💙} / {🩶} 51° {🧡} | 🌫  |
| -------- | ------------------------ | --- |
| Tue, 1/3 | 29° {💙} / {🩶} 63° {🧡} | 🥵  |
| -------- | ------------------------ | --- |
| Wed, 1/4 | 28° {💙} / {🩶} 44° {🧡} | 🌀  |
| -------- | ------------------------ | --- |
Updated Dec 28, 12:00am {🩶}
"
`;

exports[`lib/notion/update-current-weather > displays snow icon and accumulation if snow 1`] = `
"# ❄️ (0.98in) {🩶} 23° {💜}
| Wed, 12/28 | 23° {💙} / {🩶} 62° {🧡} | 🌪  |
| ---------- | ------------------------ | --- |
| Thu, 12/29 | 45° {💙} / {🩶} 66° {🧡} | ⛅️  |
| ---------- | ------------------------ | --- |
| Fri, 12/30 | 43° {💙} / {🩶} 66° {🧡} | 💨  |
| ---------- | ------------------------ | --- |
| Sat, 12/31 | 33° {💙} / {🩶} 54° {🧡} | ☔️ (85%) {🩶}  |
| ---------- | ------------------------ | -------------- |
| Sun, 1/1 | 25° {💙} / {🩶} 51° {🧡} | ❄️  |
| -------- | ------------------------ | --- |
| Mon, 1/2 | 26° {💙} / {🩶} 51° {🧡} | 🌫  |
| -------- | ------------------------ | --- |
| Tue, 1/3 | 29° {💙} / {🩶} 63° {🧡} | 🥵  |
| -------- | ------------------------ | --- |
| Wed, 1/4 | 28° {💙} / {🩶} 44° {🧡} | 🌀  |
| -------- | ------------------------ | --- |
Updated Dec 28, 12:00am {🩶}
"
`;

exports[`lib/notion/update-current-weather > updates the current weather block and daily weather table 1`] = `
"# 🥶 23° {💜}
| Wed, 12/28 | 23° {💙} / {🩶} 62° {🧡} | 🌪  |
| ---------- | ------------------------ | --- |
| Thu, 12/29 | 45° {💙} / {🩶} 66° {🧡} | ⛅️  |
| ---------- | ------------------------ | --- |
| Fri, 12/30 | 43° {💙} / {🩶} 66° {🧡} | 💨  |
| ---------- | ------------------------ | --- |
| Sat, 12/31 | 33° {💙} / {🩶} 54° {🧡} | ☔️ (85%) {🩶}  |
| ---------- | ------------------------ | -------------- |
| Sun, 1/1 | 25° {💙} / {🩶} 51° {🧡} | ❄️  |
| -------- | ------------------------ | --- |
| Mon, 1/2 | 26° {💙} / {🩶} 51° {🧡} | 🌫  |
| -------- | ------------------------ | --- |
| Tue, 1/3 | 29° {💙} / {🩶} 63° {🧡} | 🥵  |
| -------- | ------------------------ | --- |
| Wed, 1/4 | 28° {💙} / {🩶} 44° {🧡} | 🌀  |
| -------- | ------------------------ | --- |
Updated Dec 28, 12:00am {🩶}
"
`;
